File: java/src/test/java/kirillsemyonkin/lsdata/ParsingPass.java
package kirillsemyonkin.lsdata;

import java.io.ByteArrayInputStream;
import java.util.Map;

import org.junit.jupiter.api.Assertions;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import org.junit.jupiter.api.Test;

class ParsingPass {

    public static <T> void assertEquals(Class<T> type, T expected, T actual) {
        Assertions.assertEquals(expected, actual);
    }

    @Test
    void nothing() {
        var text = new ByteArrayInputStream("".getBytes());

        var lsd = assertDoesNotThrow(() -> LSD.parse(text));

        assertEquals(
                LSD.class,
                lsd,
                new LSD.Level()
        );
    }

    @Test
    void comment() {
        var text = new ByteArrayInputStream(
                """
                    # test1
                    # test2
                """.getBytes()
        );

        var lsd = assertDoesNotThrow(() -> LSD.parse(text));

        assertEquals(
                LSD.class,
                lsd,
                new LSD.Level()
        );
    }

    @Test
    void list() {
        var text = new ByteArrayInputStream("[]".getBytes());

        var lsd = assertDoesNotThrow(() -> LSD.parse(text));

        assertEquals(
                LSD.class,
                lsd,
                new LSD.List()
        );
    }

    @Test
    void level() {
        var text = new ByteArrayInputStream("{}".getBytes());

        var lsd = assertDoesNotThrow(() -> LSD.parse(text));

        assertEquals(
                LSD.class,
                lsd,
                new LSD.Level()
        );
    }

    @Test
    void basic() {
        var text = new ByteArrayInputStream(
                """
                    a 10 # comment
                    b 20
                    c a  \"test string\\nand spaces\"  b
                    d 'also\"string'
                    glued\" key\" test
                """.getBytes()
        );

        var lsd = assertDoesNotThrow(() -> LSD.parse(text));

        assertEquals(
                LSD.class,
                lsd,
                new LSD.Level(Map.of(
                        "a", new LSD.Value("10"),
                        "b", new LSD.Value("20"),
                        "c", new LSD.Value("a  test string\nand spaces  b"),
                        "d", new LSD.Value("also\"string"),
                        "glued key", new LSD.Value("test")
                ))
        );
    }

    @Test
    void nestedLevel() {
        var text = new ByteArrayInputStream(
                """
                    a {
                        a 10
                    }
                    b{}
                    c{ c 30 }
                    d{ d 40
                       2 50 }
                """.getBytes()
        );

        var lsd = assertDoesNotThrow(() -> LSD.parse(text));

        assertEquals(
                LSD.class,
                lsd,
                new LSD.Level(Map.of(
                        "a", new LSD.Level(Map.of("a", new LSD.Value("10"))),
                        "b", new LSD.Level(),
                        "c", new LSD.Level(Map.of("c", new LSD.Value("30"))),
                        "d", new LSD.Level(Map.of(
                                "d", new LSD.Value("40"),
                                "2", new LSD.Value("50")
                        ))
                ))
        );
    }

    @Test
    void nestedList() {
        var text = new ByteArrayInputStream(
                """
                    a [
                        a 10
                    ]
                    b[]
                    c[ 1 2 {} 3 4 ]
                    d[ 1 2
                       3 4 ]
                """.getBytes()
        );

        var lsd = assertDoesNotThrow(() -> LSD.parse(text));

        assertEquals(
                LSD.class,
                lsd,
                new LSD.Level(Map.of(
                        "a", new LSD.List(new LSD.Value("a 10")),
                        "b", new LSD.List(),
                        "c", new LSD.List(
                                new LSD.Value("1 2"),
                                new LSD.Level(),
                                new LSD.Value("3 4")
                        ),
                        "d", new LSD.List(
                                new LSD.Value("1 2"),
                                new LSD.Value("3 4")
                        )
                ))
        );
    }
}

File: rust/tests/parsing-pass.rs
use std::io::Cursor;

use lsdata::Level;
use lsdata::List;
use lsdata::LSD;

#[test]
fn nothing() {
    let text = Cursor::new(r#""#);

    let lsd = LSD::parse(text).unwrap();

    assert_eq!(lsd, LSD::default());
}

#[test]
fn comment() {
    let text = Cursor::new(
        r#"
            # test1
            # test2
        "#,
    );

    let lsd = LSD::parse(text).unwrap();

    assert_eq!(lsd, LSD::default());
}

#[test]
fn list() {
    let text = Cursor::new("[]");

    let lsd = LSD::parse(text).unwrap();

    assert_eq!(
        lsd,
        LSD::List(List::default())
    );
}

#[test]
fn level() {
    let text = Cursor::new("{}");

    let lsd = LSD::parse(text).unwrap();

    assert_eq!(
        lsd,
        LSD::Level(Level::default())
    );
}

#[test]
fn basic() {
    let text = Cursor::new(
        r#"
            a 10 # comment
            b 20
            c a  "test string\nand spaces"  b
            d 'also"string'
            glued" key" test
        "#,
    );

    let lsd = LSD::parse(text).unwrap();

    assert_eq!(
        lsd,
        LSD::Level(Level::from([
            (
                "a".to_string(),
                LSD::Value("10".to_string())
            ),
            (
                "b".to_string(),
                LSD::Value("20".to_string())
            ),
            (
                "c".to_string(),
                LSD::Value("a  test string\nand spaces  b".to_string())
            ),
            (
                "d".to_string(),
                LSD::Value(r#"also"string"#.to_string())
            ),
            (
                "glued key".to_string(),
                LSD::Value("test".to_string())
            ),
        ]))
    );
}

#[test]
fn nested_level() {
    let text = Cursor::new(
        r#"
            a {
                a 10
            }
            b{}
            c{ c 30 }
            d{ d 40
               2 50 }
        "#,
    );

    let lsd = LSD::parse(text).unwrap();

    assert_eq!(
        lsd,
        LSD::Level(Level::from([
            (
                "a".to_string(),
                LSD::Level(Level::from([(
                    "a".to_string(),
                    LSD::Value("10".to_string()),
                )]))
            ),
            (
                "b".to_string(),
                LSD::Level(Level::default())
            ),
            (
                "c".to_string(),
                LSD::Level(Level::from([(
                    "c".to_string(),
                    LSD::Value("30".to_string()),
                )]))
            ),
            (
                "d".to_string(),
                LSD::Level(Level::from([
                    (
                        "d".to_string(),
                        LSD::Value("40".to_string()),
                    ),
                    (
                        "2".to_string(),
                        LSD::Value("50".to_string()),
                    )
                ]))
            ),
        ]))
    );
}

#[test]
fn nested_list() {
    let text = Cursor::new(
        r#"
            a [
                a 10
            ]
            b[]
            c[ 1 2 {} 3 4 ]
            d[ 1 2
               3 4 ]
            e[ 1.2 ]
        "#,
    );

    let lsd = LSD::parse(text).unwrap();

    assert_eq!(
        lsd,
        LSD::Level(Level::from([
            (
                "a".to_string(),
                LSD::List(List::from([LSD::Value(
                    "a 10".to_string()
                )]))
            ),
            (
                "b".to_string(),
                LSD::List(List::default())
            ),
            (
                "c".to_string(),
                LSD::List(List::from([
                    LSD::Value("1 2".to_string()),
                    LSD::Level(Level::default()),
                    LSD::Value("3 4".to_string())
                ]))
            ),
            (
                "d".to_string(),
                LSD::List(List::from([
                    LSD::Value("1 2".to_string()),
                    LSD::Value("3 4".to_string())
                ]))
            ),
            (
                "e".to_string(),
                LSD::List(List::from([LSD::Value(
                    "1.2".to_string()
                )]))
            ),
        ]))
    );
}
